### ISSUE TITLE
Shuffling Causes NCCL Timeouts

### ISSUE THREAD

**Issue by schopra8** [mosaicml/streaming#876]
## Environment
- OS: Ubuntu 22.04.5 LTS (GNU/Linux 5.15.0-1048-oracle x86_64) 
- Hardware (GPU, or instance type): 4 Nodes, 8xH100
- PyTorch & CUDA Versions: PyTorch 2.5.1, CUDA 12.4
- Streaming Version: 0.9.0 w/[ this fix](https://github.com/mosaicml/streaming/issues/824)
- NCCL Timeout: 30 minutes
## To reproduce

Train model with config like -- 
```
stream_configs:
    - remote: <s3://path_to_r2_dataset>
       local: /scratch/<path_to_local_dataset_on_nvme>
       proportion: <proportion_value>
    - remote: <s3://path_to_r2_dataset>
       local: /scratch/<path_to_local_dataset_on_nvme>
       proportion: <proportion_value>
    ...
predownload: 3072 # trying to provide enough of a buffer so that we don't stall out, waiting for data during training
cache_limit: XX TB # ~60% of the entire dataset can be stored in cache on local NVME drives (without running out of space)
batch_size: 128
batching_method: `device_per_stream` # each dataset has samples with different sizes; so they have to be batched per stream
shuffle: True
``` 

## Expected behavior - No NCCL Timeouts

**When we hit "breaks" in training (i.e., evaluation every K batches OR the end of the epoch), we often get NCCL timeout errors.** Sometimes we can get through multiple epochs of the dataset -- before we hit a NCCL timeout at the end of an epoch. Other times, we get NCCL timeouts every time we have an evaluation period. When we restart training, we are able to get deeper into training, so the underlying data does not seem to be corrupted. 

**If we train with `shuffle: False`, we never get NCCL timeouts -- so we're fairly confident that the NCCL timeouts are tied to shuffling logic.**  

My best hunch (pure guesswork) is that it's taking a long time to shuffle the data (and download new shards) at these "breaks" in training -- and that drives the NCCL timeout. For example, if we are at the end of an epoch - the shuffling may require the download of A LOT more data, since we cannot hold the entire dataset in cache. Perhaps, the GPUs are not waiting (in the same way) when we restart training -- so they're able to spend a lot of time downloading and shuffling the data, without the NCCL timeouts.

## Additional context

We're not using the latest version of StreamingDataset because of the errors discussed in [this thread](https://github.com/mosaicml/streaming/issues/824), from January.

**Comment by schopra8** [comment]
Edit - Changed the name from `Shuffling Causes NCCL Timeouts` to `NCCL Timeouts from Adding More Datasets + Proportion Sampling`. 

We hit the same problem with `shuffle=False` after 2 eval periods. Digging deeper, it looks like the root cause of the problem is that the time it take to go from `eval` to `train` increases proportional to the number of datasets you have.

We started hitting NCCL timeouts, when we went from 6 datasets -> 11 datasets in training and validation. We've gotten around the issue for now by reducing the validation datasets from 11 -> 2 (while using the same number of eval batches -- 15). This shaved a few minutes over the process of going from training -> eval -> training. 

Additionally, we increased the NCCL Timeout associated with DDP training from 10 minutes -> 15 minutes. Note that this **could not** be done within `Composer`. When using `streaming-dataset` in DDP we see that the distributed training is initialized by `streaming-dataset` -- not Composer. So the `timeout` value you pass to `Trainer` is not used. To increase the NCCL timeout duration. [Rather, we had to hardcode the timeout value in a our fork of `streaming-dataset`](https://github.com/mosaicml/streaming/blob/6af3216e5608899e97f2f75be47c21d4106392fb/streaming/base/distributed.py#L116).

**Comment by schopra8** [comment]
I updated the `NCCL Timeout` to 20 minutes; it got to the end of an epoch (i.e., through several evaluation periods) -- but timed out at the end of the epoch.

Any ideas for how to tackle this issue (i.e. more datasets, significantly longer time to restart training at eval / end of epoch)?

**Comment by dawood95** [comment]
I am seeing very slow data loading with a similar setup

**Comment by schopra8** [comment]
I realized that I had `drop_last=False` in my torch Dataloader. When I set it `drop_last=True`, I see that the epoch terminates earlier than the estimated number of batches and gets stuck at the beginning of the next epoch.

```
train          Epoch   1:   99%|████████████████████████▋| 46863/47382 [13:02:45<08:34,  1.01ba/s, loss/train/total=XXX]
train          Epoch   2:    0%|                         | 0/47382 [00:00<?, ?ba/s]
```

**Comment by schopra8** [comment]
Did some more digging and it looks like this behavior is tied to race conditions under eviction -- where a thread is accessing a shard that another thread has evicted.

We altered our[ fork of 0.9.0 ](https://github.com/mosaicml/streaming/issues/824)with [this fix from 0.9.1 ](https://github.com/mosaicml/streaming/pull/795) -- but we still get these errors stochastically. The more we undersample/oversample from datasets the more this issue emerges.

@snarayan21 @XiaohanZhangCMU -- **Do ya'll have any ideas on how we might address the race condition? Or, whether some of the more recent updates have tackled this issue?** My immediate idea would be to add a lock on the "last accessed" time updates, but I assume that might significantly throttle the libraries downloads/evictions.
